python/tests/backends/test_IQM.py

```diff
@@ -0,0 +1,80 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
+import os
+import tempfile
+import time
+from multiprocessing import Process
+
+import cudaq
+import pytest
+from utils.mock_qpu.iqm.mock_iqm_cortex_cli import write_a_mock_tokens_file
+from utils.mock_qpu.iqm.mock_iqm_server import startServer
+
+# Define the port for the mock server
+port = 9100
+
+
+@pytest.fixture(scope="session", autouse=True)
+def startUpMockServer():
+    # Write a fake access tokens file
+    tmp_tokens_file = tempfile.NamedTemporaryFile(delete=False)
+    write_a_mock_tokens_file(tmp_tokens_file.name)
+
+    # Launch the Mock Server
+    p = Process(target=startServer, args=(port,))
+    p.start()
+    time.sleep(1)
+
+    # Set the targeted QPU
+    os.environ["IQM_TOKENS_FILE"] = tmp_tokens_file.name
+    cudaq.set_target(
+        "iqm", url="http://localhost:{}".format(port), **{"qpu-architecture": "Apollo"}
+    )
+
+    yield "Running the tests."
+
+    # Kill the server, remove the tokens file
+    p.terminate()
+    os.remove(tmp_tokens_file.name)
+
+
+def test_iqm_sample():
+    kernel = cudaq.make_kernel()
+    qubits = kernel.qalloc(2)
+    kernel.h(qubits[0])
+    kernel.cx(qubits[0], qubits[1])
+    kernel.mz(qubits[0])
+    kernel.mz(qubits[1])
+
+    counts = cudaq.sample(kernel)
+    assert len(counts) == 4
+    assert "00" in counts
+    assert "11" in counts
+
+    future = cudaq.sample_async(kernel)
+    counts = future.get()
+    assert len(counts) == 4
+    assert "00" in counts
+    assert "11" in counts
+
+    future = cudaq.sample_async(kernel)
+
+    futureAsString = str(future)
+
+    futureReadIn = cudaq.AsyncSampleResult(futureAsString)
+    counts = futureReadIn.get()
+    assert len(counts) == 4
+    assert "00" in counts
+    assert "11" in counts
+
```

Comment:
Would you mind copying over the observe test from one of the other backend tests in this folder?

Reply:
Copying `test_ionq_observe` as is produces:

```
.loc("<builder>":1:1): error: Invalid operation, code not lowered to IQM native gate set (ry).
loc("<builder>":1:1): error: IQM phased_rx gate expects exactly two parameters.
python3: /opt/llvm-dev/include/llvm/ADT/STLExtras.h:1309: ReferenceT llvm::detail::indexed_accessor_range_base<mlir::OperandRange, mlir::OpOperand *, mlir::Value, mlir::Value, mlir::Value>::operator[](size_t) const [DerivedT = mlir::OperandRange, BaseT = mlir::OpOperand *, T = mlir::Value, PointerT = mlir::Value, ReferenceT = mlir::Value]: Assertion `Index < size() && "invalid index for value range"' failed.
Fatal Python error: Aborted
```

Reply:
```diff
diff --git a/runtime/cudaq/platform/default/rest/RemoteRESTQPU.cpp b/runtime/cudaq/platform/default/rest/RemoteRESTQPU.cpp
index c7bf07a..584b72b 100644
--- a/runtime/cudaq/platform/default/rest/RemoteRESTQPU.cpp
+++ b/runtime/cudaq/platform/default/rest/RemoteRESTQPU.cpp
@@ -337,7 +337,7 @@ public:
             cudaq::opt::createQuakeObserveAnsatzPass(binarySymplecticForm[0]));
         if (failed(pm.run(tmpModuleOp)))
           throw std::runtime_error("Could not apply measurements to ansatz.");
-        runPassPipeline("canonicalize", tmpModuleOp);
+        runPassPipeline(passPipelineConfig, tmpModuleOp);
         modules.emplace_back(term.to_string(false), tmpModuleOp);
       }
     } else
```

Reply:
@amccaskey Yes, this helps. Also needed a change on the mock server to accept multiple circuits, I've pushed the changes. However, in order to make this test work IQM mock server needs to actually return simulated results, it doesn't do it now. I can see two options: 1. we can add simulation of iqm-json-defined circuits using qiskit or 2. wait until we finally accept QIR and use llvmlite. How urgent is to have simulation in the mock server?

Reply:
@bettinaheim could you please suggest if we should add circuits simulation to the IQM mock server or this can wait until we have QIR support?

Reply:
@kukushechkin 1. is not a good option. We also will want to have tests enabled before we pull this into main and officially call it supported. What is needed for that is pretty minimal though; you don't need a full simulator. The only purpose is to mock up results for the 1 or 2 test programs that are submitted to the mock server for the purpose of testing. It should be very easy to write a quick function that takes your json input and spits out results for a the sequence of gates. I am happy to limit what programs we submit to the mock server for now rather than taking on an additional dependency or having these tests disabled.

Reply:
@amccaskey @bettinaheim added a poor man's implementation for circuits simulation in the iqm mock server.